### PR TITLE
feat: add seq2seq forecasting training job

### PR DESCRIPTION
**Pull request by TheMichaelHu**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-aiplatform/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes b/229909845 🦕

---

Adds a `SequenceToSequencePlusForecastingTrainingJob` to training jobs. This job has the exact same signature as `AutoMLForecastingTrainingJob`, but we are creating a separate job in case the two models diverge in the future.

The logic for `AutoMLForecastingTrainingJob` has been moved to a new abstract base class `_ForecastingTrainingJob`. The only things that differ between the seq2seq and automl training jobs that extend it are the `model_type` and `training_task_definition`.

### CODE REVIEW

**Comment by TheMichaelHu** [comment]
TODO: split commit for better diff

File: google/cloud/aiplatform/training_jobs.py
Comment (geraint0923) [inline google/cloud/aiplatform/training_jobs.py:4956]: SequenceToSequence**Plus**ForecastingTrainingJob

**Comment by geraint0923** [comment]
Overall LG, just one minor comment.

**Review by geraint0923 — APPROVED** [review]
LGTM, thanks!

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [inline google/cloud/aiplatform/training_jobs.py:5316]: You can just remove this if you call the super implementation.

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [inline google/cloud/aiplatform/training_jobs.py:4605]: If you just call super, can't you just remove this __init__?

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [inline google/cloud/aiplatform/training_jobs.py:4717]: Can you just remove the `run` override and defer to super?

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [inline google/cloud/aiplatform/training_jobs.py:4952]: Just remove this method and defer to super.

File: google/cloud/aiplatform/training_jobs.py
Comment (TheMichaelHu) [reply google/cloud/aiplatform/training_jobs.py:5316]: Do you know if removing this would impact the documentation?

**Review by ivanmkc — CHANGES_REQUESTED** [review]
It looks like you can simplify both AutoMLForecastingTrainingJob and SequenceToSequencePlusForecastingTrainingJob by removing all methods, since they are the exact same as the super's implementation.

We also need to add system tests for these. See tests/system/aiplatform/test_e2e_tabular.py for an example.

**Comment by ivanmkc** [comment]
Regarding my last comment, you can simplify to:

```


class AutoMLForecastingTrainingJob(_ForecastingTrainingJob):
    _model_type = "AutoML"
    _training_task_definition = schema.training_job.definition.automl_forecasting
    _supported_training_schemas = (schema.training_job.definition.automl_forecasting,)

class SequenceToSequencePlusForecastingTrainingJob(_ForecastingTrainingJob):
    _model_type = "Seq2Seq"
    _training_task_definition = schema.training_job.definition.seq2seq_plus_forecasting
    _supported_training_schemas = (
        schema.training_job.definition.seq2seq_plus_forecasting,
    )
```

I ran the unit tests and it works fine. Otherwise, everything else looks pretty good.

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [reply google/cloud/aiplatform/training_jobs.py:5316]: I think they can follow the superclass for docstrings.
Not sure if we use any automated tools that would be affected by this.

@sasha-gitg any thoughts? I would prefer less redundant code if possible.

**Comment by TheMichaelHu** [comment]
Thanks for the review @ivanmkc! qq: how will simplifying the way you suggested impact the [documentation](https://googleapis.dev/python/aiplatform/latest/aiplatform.html?highlight=batchpredictionjob#google.cloud.aiplatform.AutoMLForecastingTrainingJob.run)? Is there a way to check that there will be no impact?

File: tests/unit/aiplatform/test_automl_forecasting_training_jobs.py
Comment (ivanmkc) [inline tests/unit/aiplatform/test_automl_forecasting_training_jobs.py:274]: Move this above sync for consistency's sake.

Comment (TheMichaelHu) [reply tests/unit/aiplatform/test_automl_forecasting_training_jobs.py:274]: Can you elaborate why `training_job` should be above `sync`? I wanted to keep it aligned with the order `parameterize` was called, so should the order of the calls to `parameterize` be swapped too?

**Comment by TheMichaelHu** [comment]
> We also need to add system tests for these. See tests/system/aiplatform/test_e2e_tabular.py for an example.

Should this occur in a separate PR since we never had e2e tests for Vertex Forecasting in the first place? If so, I can make a GitHub issue to track it and assign to myself.

File: google/cloud/aiplatform/training_jobs.py
Comment (TheMichaelHu) [reply google/cloud/aiplatform/training_jobs.py:5316]: `nox -s docs` isn't creating documentation for `run()` as expected, since it's now part of the parent class. Looking into ways to force a child class to document a method from the parent class.

Comment (TheMichaelHu) [reply google/cloud/aiplatform/training_jobs.py]: Removed `evaluated_data_items_bigquery_uri`, moving discussion to the `run` method which I kept.

File: google/cloud/aiplatform/training_jobs.py
Comment (TheMichaelHu) [reply google/cloud/aiplatform/training_jobs.py]: If we remove this, Sphinx will not generate documentation for the `run()` function. It can pull the docstring from the parent class (tested with `nox -s docs`), but I couldn't find a way to make it document the function signature other than by writing it out manually.

File: tests/unit/aiplatform/test_automl_forecasting_training_jobs.py
Comment (ivanmkc) [inline tests/unit/aiplatform/test_automl_forecasting_training_jobs.py:516]: training_jobs.AutoMLForecastingTrainingJob and training_jobs.SequenceToSequencePlusForecastingTrainingJob will not necessarily share the same API forever. Prefer splitting into two separate tests.

File: tests/unit/aiplatform/test_automl_forecasting_training_jobs.py
Comment (ivanmkc) [inline tests/unit/aiplatform/test_automl_forecasting_training_jobs.py:439]: Ditto. Split into two tests.

File: tests/unit/aiplatform/test_automl_forecasting_training_jobs.py
Comment (ivanmkc) [inline tests/unit/aiplatform/test_automl_forecasting_training_jobs.py:374]: Ditto.

File: tests/unit/aiplatform/test_automl_forecasting_training_jobs.py
Comment (ivanmkc) [reply tests/unit/aiplatform/test_automl_forecasting_training_jobs.py:274]: Okay I see, feel free to leave it as is.

File: tests/unit/aiplatform/test_automl_forecasting_training_jobs.py
Comment (ivanmkc) [reply tests/unit/aiplatform/test_automl_forecasting_training_jobs.py:516]: I'm torn. I'm also open to deferring this until it's needed. Feel free to decide.

Comment (TheMichaelHu) [reply tests/unit/aiplatform/test_automl_forecasting_training_jobs.py:516]: Let's leave it as-is for now. The intention is for Seq2Seq to share the same API as AutoML, but we're leaving the door open for implementation differences on the SDK side by splitting into two classes instead of just having a `model_type` argument in the `AutoMLForecastingTrainingJob`. At least that's my understanding.

Comment (ivanmkc) [reply tests/unit/aiplatform/test_automl_forecasting_training_jobs.py:516]: Sure, we can revisit later. Thanks.

File: google/cloud/aiplatform/training_jobs.py
Comment (ivanmkc) [reply google/cloud/aiplatform/training_jobs.py]: Hm, that's quite annoying. Let me ask the SDK team for more solutions.

**Review by ivanmkc — APPROVED** [review]
LGTM if system tests pass.

**Pull request merged**
